graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/method/MethodDataFetcherFactory.kt

```diff
@@ -29,12 +34,53 @@ import java.lang.reflect.Method
  * Resolving of method arguments is handled by the supplied [argument resolvers][ArgumentResolver].
  */
 class MethodDataFetcherFactory(
-    argumentResolvers: List<ArgumentResolver>,
+    private val argumentResolvers: List<ArgumentResolver>,
     private val parameterNameDiscoverer: ParameterNameDiscoverer = DefaultParameterNameDiscoverer()
 ) {
 
     private val resolvers = ArgumentResolverComposite(argumentResolvers)
 
+    fun checkInputArgumentsAreValid(method: Method, argumentNames: Set<String>) {
```

Comment:
I'm not sure that this method needs to live in `MethodDataFetcherFactory`. I think some of the checks themselves are basically a no-op in practice, since we're filtering parameters that are annotated with `InputArgument`, and the `InputArgumentResolver` matches any method parameter with this annotation.  I think the `SchemaProvider` could simply perform the check directly by just collecting field names from the annotated parameters.

Reply:
Let me reply to both of your comments in one go as they are related.

> I'm not sure that this method needs to live in MethodDataFetcherFactory. I think some of the checks themselves are basically a no-op in practice, since we're filtering parameters that are annotated with InputArgument, and the InputArgumentResolver matches any method parameter with this annotation. I think the SchemaProvider could simply perform the check directly by just collecting field names from the annotated parameters.

I went back and forth on this, between adding the logic in the `Factory` or in the `SchemaProvider` and I opted for the  `Factory` for the following reasons:

 - if we were to add this in the schema provider, we would need to inject the resolvers that are also injected in the `Factory`. The `ArgumentResolverComposite` would also possibly need to be replicated to match what is being done at runtime. This would make the argument resolution & validation logic more brittle as it would live in multiple places.

 - The argument resolvers list is auto-wired and DGS owners could be injecting their own list of ArgumentResolvers. If that's the case our validation should run if and only if our specific `InputArgument` resolver would be selected by the composite. The way to do that without leaking the implementation elsewhere was to place it in the `Factory`. 
 
I also contemplated the idea of introducing a new `ArgumentResolver` interface method so that each argument resolver could optionally perform their own validation, but I felt `ArgumentResolver` is about resolution not validation and so I abandoned the idea. We could introduce yet another interface only for validation but I thought it would be overkill for now, at least. Happy to revisit.

> Kind of related to my other comment. Is there a test that actually exercises this case? I don't know that it's possible, since we are only checking parameters that are annotated with @InputArgument, so basically by definition the resolver would always be InputArgumentResolver. IMO it isn't necessary to check the resolvers, we could simply validate that the field name from the annotation matches something from the schema, and I don't know that it needs to live here.

Great catch, I thought I added one but I think I got confused on the behavior of the `FallbackEnvironmentArgumentResolver` initially and forgot to go back and update. The idea is, if the `FallbackEnvironmentArgumentResolver` is selected, the resolved argument name will be the method parameter name and we can check whether there is a matching argument on the schema. I added a couple of tests and fixed the factory logic to look at **all** the method parameters instead of the ones annotated with `@InputArgument` in my last [commit](https://github.com/Netflix/dgs-framework/pull/1565/commits/fbb5ebefd6a6ef195958648dade9eeb7c983a7af). 

Now the problem is the following test is failing [`An unknown argument should be null, and not result in an error`](https://github.com/Netflix/dgs-framework/blob/e7e868159cb9af22a3d340a737db68b768bfec6b/graphql-dgs/src/test/kotlin/com/netflix/graphql/dgs/internal/InputArgumentTest.kt#L1034) which leads me to believe that the behavior of having a `null` value passed in when nothing matches was intended. We can definitely argue about whether it was a good decision, but at this point changing this behavior would be effectively breaking. 

To conclude, I'm leaning towards only running validation for fields annotated with `@InputArgument`, the same way I'm doing it now just we won't look at `FallbackEnvironmentArgumentResolver` and keep the behavior of a fallback as a best effort, last resort resolver that doesn't get any schema build time validation benefit. Thoughts?

Reply:
@tinnou and I just had a quick discussion and decided on 
(1) Removing the logic for checking the fallback resolver altogether, and that would allow anything without an @InputArgument to go through 
(2) Introduce a helper that does line 40 and 51 in the MethodDataFetcherFactory. This will allow other classes to determine the resolver used for a given method
(3) Move the rest of the validation logic to the schema provider, which will use this new helper.  It would be good I think to have the validation logic use the InputArgumentResolver to avoid leaking/reimplementing details of how @InputArgument annotation works - specifically since you can have something like `@InputArgument(name=something) somethingElse: String` 

Does that seem reasonable?